[stdlib] Enhance Linked List

owenhilyard opened the pull request:
Adds a `LinkedList` type which implements most of the API surface of `List` (also grabbing the tests). This should help with filling in basic data structures in the stdlib.

No iterator because I ran into some nasty issues with origins and I'm not sure how far I can push casting them before I get to UB. 

There's also an instance of `list.__getitem__(0).__getitem__(1).value = "Mojo"` in test_linked_list.mojo since I couldn't find another way to avoid the copy. If someone has a solution please let me know.

owenhilyard committed:
[stdlib] Expand LinkedList API and add additional testing

Merges my earlier work with the one which appeared in the stdlib,
expanding the API and test coverage.

Also makes a few of the test_utils counters `Writable` to help them
conform to more APIs.

Signed-off-by: Owen Hilyard <hilyard.owen@gmail.com>

JoeLoser reviewed (commented):
Just did a quick pass of this, looking good!  Left some comments and @lsh will also take a look.

Thanks for working on improving this a ton from the basic implementation that was started prior to your work!  Sorry that we already had a PR in-flight starting the linked list type.

owenhilyard committed:
[stdlib] Extended Linked List cleanup and fixups

Minor tweaks, mostly consisting of removing unnecessary things like some
leftover debug prints and explicit type parameters. Also adds time
complexity information to all public methods of `LinkedList`.

Signed-off-by: Owen Hilyard <hilyard.owen@gmail.com>
